### ISSUE TITLE
Caching of builds broken in nx > 16.1.4

### ISSUE THREAD

**Issue by atsjo** [nrwl/nx#17081]
### Current Behavior

Cached library builds is not reused...

### Expected Behavior

Cached library builds is reused in new builds, as in 16.1.4.

### Steps to Reproduce

1. upgrade to 16.2.1
2. run yarn nx reset
3. build app depending on library
4. build library (library builds again, even if nothing changed)

In 16.1.4, the last build would reuse cached build...

### Nx Report

```shell
Node   : 18.16.0
   OS     : win32 x64
   yarn   : 4.0.0-rc.44
   Hasher : Native
   
   nx                 : 16.2.1
   @nx/js             : 16.2.1
   @nx/jest           : 16.2.1
   @nx/linter         : 16.2.1
   @nx/workspace      : 16.2.1
   @nx/angular        : 16.2.1
   @nx/cypress        : 16.2.1
   @nx/devkit         : 16.2.1
   @nx/esbuild        : 16.2.1
   @nx/eslint-plugin  : 16.2.1
   @nrwl/tao          : 16.2.1
   @nx/webpack        : 16.2.1
   typescript         : 5.0.4
   ---------------------------------------
   Community plugins:
   angular-calendar : 0.31.0
```

### Operating System

- [ ] macOS
- [ ] Linux
- [X] Windows
- [ ] Other (Please specify)

**Comment by atsjo** [comment]
seems to work fine if not running yarn nx reset after upgrading...

**Comment by atsjo** [comment]
I suspect this https://github.com/nrwl/nx/pull/17139 is a fix for this issue

**Comment by atsjo** [comment]
still the same with 16.3.0... Am I the only one experiencing this? must be something special about my workspace...

If I run yarn nx run-many --target=build and then build one of the individual projects after that, it re-builds, not using the cache....

**Comment by atsjo** [comment]
16.3.2 seems to fix this, even if I can't find anything referring this issue in the changelog... so closing the issue

**Comment by atsjo** [comment]
Unfortunately this was not fixed... Sometimes it works, and sometimes not...

Now I have also migrated my repo to pnpm, and I'm experiencing the same issues as I did with yarn...

**Comment by atsjo** [comment]
I think I managed to track down the problem...

I was bundling firebase functions via esbuild in a dist folder below the project (not dist in repo root), and generated a package.json with external deps there. When this happened many cached project build was invalidated. I avoided this by adding the project dist folder to .nxignore, and then everything worked as expected.

Then I re-wrote the firebase functions setup to bundle below the root dist folder instead, which is a better solution anyways, and aligned with the standard way nx projects are configured. That also works as expected...

So the problem seems to be that in 16.2.0 something changed that invalidates caches if a project builds and writes output with a package.json below the project dist folder (which was probably auto excluded from nx before).

Closing this as noboby else seems to have this problem, and it works as expected if building to the standard root dist folder, as most projects do...
